### PR TITLE
auth: Check provider account creation date

### PR DESCRIPTION
**Pull request by evict**
This PR will disallow accounts signing up/in through GitLab or GitHub, that are newer than 14 days.

## Test plan
Tested changes locally with both GitHub and GitLab accounts.

### CODE REVIEW

File: internal/extsvc/gitlab/users.go [hunk 1]
```diff
@@ -19,6 +20,11 @@ type User struct {
 	Identities []Identity `json:"identities"`
 }
 
+type AuthUser struct {
```
Comment (evict) [inline internal/extsvc/gitlab/users.go:23]: We are using `User` as an `Author`, but these objects are not exactly the same. Creating an `AuthUser` so we can add more fields.

Comment (michaellzc) [reply internal/extsvc/gitlab/users.go:23]: maybe just update their tests? it seems straightforward to do so

```
--- FAIL: TestGitLabSource_ChangesetSource/LoadChangeset (0.02s)
--
  | --- FAIL: TestGitLabSource_ChangesetSource/LoadChangeset/integration (0.02s)
  | --- FAIL: TestGitLabSource_ChangesetSource/LoadChangeset/integration/GitlabSource_LoadChangeset_found (0.01s)
  | gitlab_test.go:616: (-want, +got):
  | (
  | """
  | ... // 28 identical lines
  | "avatar_url": "https://gitlab.com/uploads/-/system/user/avatar/3294801/avatar.png",
  | "web_url": "https://gitlab.com/ryan-blunden",
  | + 	   "created_at": "0001-01-01T00:00:00Z",
  | "identities": null
  | },
  | ... // 9 identical lines
  | """
  | )
```

**Review by michaellzc — APPROVED** [review]

File: internal/extsvc/gitlab/users.go (hunk 1 above)
Comment (evict) [reply internal/extsvc/gitlab/users.go:23]: Yes, but then the test files are not a real representation of the server response.

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go [hunk 2]
```diff
@@ -42,6 +44,15 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
 		return nil, "Could not read GitLab user from callback request.", errors.Wrap(err, "could not read user from context")
 	}
 
+	if envvar.SourcegraphDotComMode() {
```
[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go:47]: 😭

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go [hunk 3]
```diff
@@ -42,6 +44,15 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
...
+
+		twoWeeksAgo := time.Now().Add(-13 * 24 * time.Hour)
```
[the remark this answers is not in the document]
Comment (evict) [reply enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go]: made it an part of the experimental features site-config 👍

File: enterprise/cmd/frontend/internal/auth/githuboauth/session.go [hunk 4]
```diff
@@ -53,6 +57,15 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
 		}
 		return nil, "Could not read GitHub user from callback request.", err
 	}
+	exp := conf.ExperimentalFeatures()
+	if envvar.SourcegraphDotComMode() && exp.MinExtAccountAge > 0 {
+
+		twoWeeksAgo := time.Now().Add(time.Duration(exp.MinExtAccountAge) * 24 * time.Hour)
+
+		if ghUser.CreatedAt.After(twoWeeksAgo) {
+			return nil, "User account was created less than 14 days ago", errors.New("user account too new")
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/githuboauth/session.go:66]: I think you forgot the negative here? Right now the time would look from future to me.

Also The variable name and the error message is no longer accurate now.

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go [hunk 5]
```diff
@@ -42,6 +45,17 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
...
 
+	exp := conf.ExperimentalFeatures()
+
+	if envvar.SourcegraphDotComMode() && exp.MinExtAccountAge > 0 {
+
+		twoWeeksAgo := time.Now().Add(time.Duration(exp.MinExtAccountAge) * 24 * time.Hour)
+
+		if gUser.CreatedAt.After(twoWeeksAgo) {
+			return nil, "User account was created less than 14 days ago", errors.New("user account too new")
+		}
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go:56]: Same as the gitlab one, `twoWeeksAgo` is "two weeks later" at the moment

File: internal/extsvc/gitlab/users.go (hunk 1 above)
Comment (unknwon) [reply internal/extsvc/gitlab/users.go:23]: Could you add some docstring as context as why `AuthUser` need to exist? It is very easy for someone else look at this code and thought they're just so similar and merge them, update the tests. 😂

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go (hunk 5 above)
Comment (evict) [reply enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go]: Good catch!

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go [hunk 6]
```diff
@@ -42,6 +45,17 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
 		return nil, "Could not read GitLab user from callback request.", errors.Wrap(err, "could not read user from context")
 	}
 
+	exp := conf.ExperimentalFeatures()
+
+	if envvar.SourcegraphDotComMode() && exp.MinExtAccountAge > 0 {
```
Comment (unknwon) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go:50]: nit: we don't need dotcom check now, it's a config 🎉 (same for github one)

**Review by unknwon — APPROVED** [review]
Thank you!

File: enterprise/cmd/frontend/internal/auth/githuboauth/session.go [hunk 7]
```diff
@@ -53,6 +56,15 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
 		}
 		return nil, "Could not read GitHub user from callback request.", err
 	}
+	exp := conf.ExperimentalFeatures()
+	if exp.MinExtAccountAge > 0 {
+
+		twoWeeksAgo := time.Now().Add(time.Duration(-exp.MinExtAccountAge) * 24 * time.Hour)
```
Comment (mrnugget) [inline enterprise/cmd/frontend/internal/auth/githuboauth/session.go:62]: It's not two weeks ago though, right? How about `earliestValidCreationDate` or something?

File: schema/site.schema.json [hunk 8]
```diff
@@ -130,6 +130,11 @@
           "type": "integer",
           "default": 500
         },
+        "minExtAccountAge": {
```
Comment (eseliger) [inline schema/site.schema.json:133]: can we nest this under `dotcom`? Also, I would call it `minimumExternalAccountAge` probably

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go [hunk 9]
```diff
@@ -42,6 +44,17 @@ func (s *sessionIssuerHelper) GetOrCreateUser(ctx context.Context, token *oauth2
 		return nil, "Could not read GitLab user from callback request.", errors.Wrap(err, "could not read user from context")
 	}
 
+	exp := conf.ExperimentalFeatures()
+
+	if exp.MinExtAccountAge > 0 {
+
+		twoWeeksAgo := time.Now().Add(time.Duration(-exp.MinExtAccountAge) * 24 * time.Hour)
```
Comment (mrnugget) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go:51]: Same as above

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go [hunk 10]
```diff
@@ -232,7 +232,7 @@ func TestSessionIssuerHelper_GetOrCreateUser(t *testing.T) {
 			inputs: []input{{
 				description: "glUser, in allowed subgroup  -> session created",
 				allowGroups: []string{"group1/subgroup2"},
-				glUser: &gitlab.User{
+				glUser: &gitlab.AuthUser{
```
Comment (mrnugget) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/session_test.go:235]: Can't we add a test for the new behaviour?

File: enterprise/cmd/frontend/internal/auth/githuboauth/session.go (hunk 7 above)
Comment (eseliger) [inline enterprise/cmd/frontend/internal/auth/githuboauth/session.go:62]: this is not actually two weeks ago, is it?

File: enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go (hunk 9 above)
Comment (eseliger) [inline enterprise/cmd/frontend/internal/auth/gitlaboauth/session.go:51]: this is not actually two weeks ago, is it? 

same for the error message below

**Comment by mrnugget** [comment]
@unknwon what about the reviews Erik and I left? Did you see them?

**Comment by unknwon** [comment]
> @unknwon what about the reviews Erik and I left? Did you see them?

@mrnugget - My bad, went offline after hitting auto merge. @evict should have addressed them in https://github.com/sourcegraph/sourcegraph/pull/56332, LMK if there is anything else missed!

**Pull request merged**